narwhals/utils.py

```diff
@@ -297,14 +302,48 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
         4  1
         5  2
     """
+
     df_any = cast(Any, df)
     native_frame = to_native(df_any)
-    if is_pandas_like_dataframe(native_frame):
-        return df_any._from_compliant_dataframe(  # type: ignore[no-any-return]
-            df_any._compliant_frame._from_native_frame(
-                native_frame.set_index(column_names)
+
+    if column_names is not None and index is not None:
+        msg = "Only one of `column_names` or `index` should be provided"
+        raise ValueError(msg)
+
+    if not column_names and not index:
+        msg = "Either `column_names` or `index` should be provided"
+        raise ValueError(msg)
+
+    if column_names is not None:
+        if is_pandas_like_dataframe(native_frame):
+            return df_any._from_compliant_dataframe(  # type: ignore[no-any-return]
+                df_any._compliant_frame._from_native_frame(
+                    native_frame.set_index(column_names)
+                )
             )
-        )
+        elif is_pandas_like_series(native_frame):
+            msg = "Cannot set index using column names on a Series"
+            raise ValueError(msg)
+
+    if index is not None:  # pragma: no cover
+        from narwhals.series import Series
+
+        if _is_iterable(index):
+            index = [idx.to_native() if isinstance(idx, Series) else idx for idx in index]
+        if isinstance(index, Series):
+            index = index.to_native()
+
+        if is_pandas_like_dataframe(native_frame):
+            return df_any._from_compliant_dataframe(  # type: ignore[no-any-return]
+                df_any._compliant_frame._from_native_frame(native_frame.set_index(index))
+            )
+
+        elif is_pandas_like_series(native_frame):
+            native_frame.index = index
```

Comment:
@MarcoGorelli pros and cons of setting the index like this vs using [`Series.set_axis`](https://pandas.pydata.org/docs/reference/api/pandas.Series.set_axis.html)?